skbio/metadata/_mixin.py

```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+        self.interval_metadata = interval_metadata
+
+    @property
+    @experimental(as_of="0.5.0-dev")
+    def interval_metadata(self):
+        '''``IntervalMetadata`` object containing info about interval features.
+
+        >>> from skbio import DNA
+        >>> from skbio.metadata import IntervalMetadata
+        >>> im = IntervalMetadata()
+        >>> intvl = im.add([(0, 3)], metadata={'gene': 'sagA'})
+        >>> seq = DNA('ACGT', interval_metadata=im)
+        >>> seq
+        DNA
+        --------------------------
+        Interval metadata:
+            1 interval features
+        Stats:
+            length: 4
+            has gaps: False
+            has degenerates: False
+            has definites: True
+            GC-content: 50.00%
+        --------------------------
+        0 ACGT
+        '''
+        if self._interval_metadata is None:
+            # Not using setter to avoid copy.
+            self._interval_metadata = IntervalMetadata()
+        return self._interval_metadata
+
+    @interval_metadata.setter
+    def interval_metadata(self, interval_metadata):
+        if interval_metadata is None:
+            interval_metadata = IntervalMetadata()
+        if isinstance(interval_metadata, IntervalMetadata):
+            # copy all the data to the mixin
+            self._interval_metadata = copy.deepcopy(interval_metadata)
+        else:
+            raise TypeError('You must provide `IntervalMetadata` object.')
+
+    @interval_metadata.deleter
+    def interval_metadata(self):
+        self._interval_metadata = None
+
+    @experimental(as_of="0.5.0-dev")
+    def has_interval_metadata(self):
+        """Determine if the object has interval metadata.
+
+        An object has interval metadata if its ``interval_metadata``
+        has at least one ```Interval`` objects.
+
+        Returns
+        -------
+        bool
+            Indicates whether the object has interval metadata.
+
+        Examples
+        --------
+        .. note:: scikit-bio objects with interval metadata share a common
+           interface for accessing and manipulating their interval metadata.
+           The following examples use scikit-bio's ``DNA`` class for
+           demonstration. These examples apply to all other
+           scikit-bio objects storing interval metadata.
+
+        >>> from skbio import DNA
+        >>> seq = DNA('ACGT')
+        >>> seq.has_interval_metadata()
+        False
+        >>> from skbio.metadata import IntervalMetadata
+        >>> im = IntervalMetadata()
+        >>> seq = DNA('ACGT', interval_metadata=im)
+        >>> seq.has_interval_metadata()
+        False
+        >>> intvl = im.add([(0, 1)], metadata={'gene': 'sagA'})
+        >>> seq = DNA('ACGT', interval_metadata=im)
+        >>> seq.has_interval_metadata()
+        True
+        """
+        return (self._interval_metadata is not None and
+                len(self.interval_metadata._intervals) > 0)
```

Comment:
It'd be more Pythonic, but we had originally deferred defining `__len__`, `__iter__`, `__getitem__`, `__setitem__`, etc., which all must agree on what "length" means for `IntervalMetadata`. We weren't able to come to a conclusion in the RFC because we weren't sure how/if `Sequence.__getitem__` would use `IntervalMetadata.__getitem__`. `__getitem__`'s behavior will determine the behavior of `__len__`, `__iter__`, etc. so we can't define those yet.

Defining these methods depends on whether we see `IntervalMetadata` as a Python "sequence" or a "mapping". For details on sequence and mapping protocols, see [[1]](https://docs.python.org/3/reference/datamodel.html#emulating-container-types) and [[2]](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes). I think we should defer defining these methods until we have a clearer idea of what makes the most sense / what is useful for them to do (and that'll require some discussion). To move this forward, I think we should define an attribute with a clear name that isn't tied to the magic methods listed above. That way we are free to define those later in a backwards-compatible way. I like `num_interval_features` or `interval_feature_count` since they're very descriptive names, and that's how they're described in the repr and `has_interval_metadata`. What do you think?